text/0000-infered-types.md

```diff
@@ -0,0 +1,174 @@
+- Feature Name: Inferred Types
+- Start Date: 2023-06-06
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+
+# Summary
+[summary]: #summary
+
+This RFC introduces a feature allowing the base type of enums and structs to be inferred in certain contexts including but not limited to function calls. The syntax is as follows `_::Variant`.
+
+
+# Motivation
+[motivation]: #motivation
+
+Using large libraries usually requires developers to import large amounts of traits, structures, and enums. To just call a function on an implementation can take upwards of 3 imports. One way developers have solved this is by importing everything from specific modules. Importing everything has its own problems like trying to guess where imports are actually from. Developers need a low compromise solution to solve this problem and that comes in the form of inferred types.
+
+Swift has had a system to infer types since around 2014. Its system has been praised by many developers around the world. It’s system is as follows:
+```swift
+enum EnumExample {
+   case variant
+   case variant2
+}
+
+
+example(data: .variant);
+```
+
+This RFC’s intent was to create something similar to the system already tried and tested in swift. Additionally, the underscore is already used to imply type’s lifetimes so it runs consistent with the rust theme.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+Implied types basically replace the path to a type with `_`. They can be used in places where strong typing exists. This includes function calls and function returns. You can think about the `_` as expanding into the type’s name.
+
+Function calls (struct):
+```rust
+fn my_function(data: MyStruct) { /* ... */ }
+
+// my_function(MyStruct {
+//     value: 1
+// });
+my_function(_ {
+   value: 1
+});
```

Comment:
Does this mean it will become a BC break to change the signature to `fn my_function(data: impl SomeTrait)` where `MyStruct: SomeTrait`?

Reply:
Yes because sudenly the type is not specific enough.

Reply:
That has _always_ been a breaking change, because you could write

```rs
my_function(something.into());
```

and once `my_function` is changed to accept `impl SomeTrait` instead of a concrete type, this fails to typecheck!

Reply:
Yeah, but it is rare enough that cargo [documents it as minor change](https://doc.rust-lang.org/cargo/reference/semver.html#minor-generalizing-a-function-to-use-generics-supporting-original-type).

This feature would significantly change that, likely making the change always be a major one.